server/sonar-server/src/main/java/org/sonar/server/computation/measure/MeasureRepositoryImpl.java

```diff
@@ -19,39 +19,81 @@
  */
 package org.sonar.server.computation.measure;
 
+import com.google.common.base.Function;
 import com.google.common.base.Optional;
 import com.google.common.base.Predicate;
+import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Iterables;
+import com.google.common.collect.Maps;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
 import javax.annotation.Nonnull;
-import org.sonar.api.measures.Metric;
+import javax.annotation.Nullable;
 import org.sonar.batch.protocol.output.BatchReport;
 import org.sonar.core.measure.db.MeasureDto;
 import org.sonar.core.persistence.DbSession;
 import org.sonar.server.computation.batch.BatchReportReader;
 import org.sonar.server.computation.component.Component;
+import org.sonar.server.computation.metric.Metric;
+import org.sonar.server.computation.metric.MetricRepository;
 import org.sonar.server.db.DbClient;
 
+import static com.google.common.collect.FluentIterable.from;
+import static java.util.Objects.requireNonNull;
+
 public class MeasureRepositoryImpl implements MeasureRepository {
   private final DbClient dbClient;
   private final BatchReportReader reportReader;
+  private final MeasureDtoToMeasure measureDtoToMeasure;
+  private final Function<BatchReport.Measure, Measure> batchMeasureToMeasureFunction;
+  private final BatchMeasureToMeasure batchMeasureToMeasure;
+  private final Map<Integer, Map<String, Measure>> measures = new HashMap<>();
 
-  public MeasureRepositoryImpl(DbClient dbClient, BatchReportReader reportReader) {
+  public MeasureRepositoryImpl(DbClient dbClient, BatchReportReader reportReader,
+    MeasureDtoToMeasure measureDtoToMeasure, final BatchMeasureToMeasure batchMeasureToMeasure,
+    final MetricRepository metricRepository) {
     this.dbClient = dbClient;
     this.reportReader = reportReader;
+    this.measureDtoToMeasure = measureDtoToMeasure;
+    this.batchMeasureToMeasure = batchMeasureToMeasure;
```

Comment:
>What's the value for injecting this object instead of using a static method like BatchMeasureToMeasure.toMeasure() ?

none, removing the constructor parameters

> Note that QualityGateServiceImpl is slightly different: the field conditionDtoToBean is instantiated directly in the constructor and is not injected by container.

`conditionDtoToBean` is a `Function`, it can not be injected. But it depends on the injected `MetricRepository` (so, must be stored in a property to avoid useless multiple instanciations). Since the `MetricRepository` is used only in this `Function` we do not create a useless `MetricRepository` property and the `Function` is created in the constructor